### PR TITLE
Yet more Docker improvements

### PR DESCRIPTION
**Pull request by JayH5**
### Change summary:
* Clean up package indexes
* Build everything with OpenSSL 1.1.0 (HAProxy 1.7 & recent `cryptography` support it)
* Add `ca-certificates` as a runtime dependency so HTTPS certificate verification works
* Install tini in a more complicated but cleaner way
* Merge `build-haproxy.sh` into the Dockerfile so everything about the build process can be seen in one place, use build commands similar to those used in official Docker images
* Clean up `make` commands a bit
* Fix invalid HAProxy `CPU` build option
* Update `tini` to 0.13.2

### Before:
```
REPOSITORY               TAG                 IMAGE ID            CREATED             SIZE
mesosphere/marathon-lb   git.e804f5d         9123d3835779        12 seconds ago      196.1 MB
```

```
HA-Proxy version 1.7.1 2016/12/13
Copyright 2000-2016 Willy Tarreau <willy@haproxy.org>

Build options :
  TARGET  = linux2628
  CPU     = x86_64
  CC      = gcc
  CFLAGS  = -g -fno-strict-aliasing -Wdeclaration-after-statement
  OPTIONS = USE_ZLIB=1 USE_REGPARM=1 USE_OPENSSL=1 USE_LUA=1 USE_STATIC_PCRE=1 USE_PCRE_JIT=1

Default settings :
  maxconn = 2000, bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Encrypted password support via crypt(3): yes
Built with zlib version : 1.2.8
Running on zlib version : 1.2.8
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with OpenSSL version : OpenSSL 1.0.2j  26 Sep 2016
Running on OpenSSL version : OpenSSL 1.0.2j  26 Sep 2016
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports prefer-server-ciphers : yes
Built with PCRE version : 8.39 2016-06-14
Running on PCRE version : 8.39 2016-06-14
PCRE library supports JIT : yes
Built with Lua version : Lua 5.3.3
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available filters :
	[COMP] compression
	[TRACE] trace
	[SPOE] spoe

```

### After:
```
REPOSITORY               TAG                 IMAGE ID            CREATED             SIZE
mesosphere/marathon-lb   git.36bcb7c         d49fcda4046c        21 seconds ago      178.5 MB
```

```
HA-Proxy version 1.7.1 2016/12/13
Copyright 2000-2016 Willy Tarreau <willy@haproxy.org>

Build options :
  TARGET  = linux2628
  CPU     = generic
  CC      = gcc
  CFLAGS  = -m64 -march=x86-64 -O2 -g -fno-strict-aliasing -Wdeclaration-after-statement
  OPTIONS = USE_ZLIB=1 USE_REGPARM=1 USE_OPENSSL=1 USE_LUA=1 USE_STATIC_PCRE=1 USE_PCRE_JIT=1

Default settings :
  maxconn = 2000, bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Encrypted password support via crypt(3): yes
Built with zlib version : 1.2.8
Running on zlib version : 1.2.8
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with OpenSSL version : OpenSSL 1.1.0c  10 Nov 2016
Running on OpenSSL version : OpenSSL 1.1.0c  10 Nov 2016
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports prefer-server-ciphers : yes
Built with PCRE version : 8.39 2016-06-14
Running on PCRE version : 8.39 2016-06-14
PCRE library supports JIT : yes
Built with Lua version : Lua 5.3.3
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available filters :
	[COMP] compression
	[TRACE] trace
	[SPOE] spoe

```

### Notes:
* This does make the Dockerfile more complex by ditching the separate build script. I think there are advantages to this: it makes it very easy to see how everything is installed. It's clear what the build dependencies in the Dockerfile are for. The build steps for e.g. HAProxy look very similar to those in the [official HAProxy Dockerfile](https://github.com/docker-library/haproxy/blob/master/1.7/Dockerfile) now too.
* OpenSSL 1.1.0 adds Poly1305/ChaCha20 cipher support so that could be re-enabled since the #373 changes (although I wouldn't necessarily recommend that they are made the first choice of ciphersuite).

### CODE REVIEW

**Comment by brndnmtthws** [comment]
test this please

File: Dockerfile [hunk 1]
```diff
@@ -2,20 +2,39 @@ FROM debian:stretch
...
+    && gpg --keyserver ha.pool.sks-keyservers.net --recv-keys "$TINI_GPG_KEY" \
+    && gpg --batch --verify tini.asc tini \
+    && rm -r "$GNUPGHOME" tini.asc \
+    && mv tini /usr/bin/tini \
+    && chmod +x /usr/bin/tini \
+    && tini -- true \
+    && apt-get purge -y --auto-remove dirmngr gpg wget
+
+
+ENV LUA_VERSION=5.3.3 \
```
Comment (brndnmtthws) [inline Dockerfile]: Might as well bump this to 5.3.4 while you're at it.

Comment (JayH5) [inline Dockerfile]: You want to use the 5.3.4 release candidate?

Comment (brndnmtthws) [inline Dockerfile]: Woops, I thought it was stable. Nevermind :)

**Pull request merged**